### PR TITLE
#35 geocode interpretation

### PR DESCRIPTION
**Pull request by asturcon**
Closes the #35 

Technical Debt:
- I had to duplicate some code in the gbif and labs modules to mock the WS calls. We should either find another way to mock this calls when testing PTransforms(preferred) or find a way to share test resources across the modules.
- I am passing to the Trace class only the first term of the interpretation, but maybe Trace should receive a list of terms.
- I ignored the TaxonomyInterpretationPipelineTest. I think we could remove this test and the pipeline. I need to review it more carefully.

### CODE REVIEW

File: core/src/main/java/org/gbif/pipelines/core/interpretation/LocationInterpreter.java [hunk 1]
```diff
@@ -212,17 +228,27 @@ static LocationInterpreter interpretCoordinateUncertaintyInMeters(Location locat
...
-        Double result = parseResult.orElse(null);
-        if (result != null && result >= COORDINATE_PRECISION_LOWER_BOUND && result <= COORDINATE_PRECISION_UPPER_BOUND) {
-          locationRecord.setCoordinatePrecision(result);
-        } else {
-          interpretation.withValidation(Trace.of(DwcTerm.coordinatePrecision.name(), IssueType.COORDINATE_PRECISION_INVALID));
-        }
-        return interpretation;
-      });
+    return (ExtendedRecord extendedRecord) -> SimpleTypeParser.parseDouble(extendedRecord,
+                                                                           DwcTerm.coordinatePrecision,
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/interpretation/LocationInterpreter.java]: indentation is different to others lambda functions

File: core/src/main/java/org/gbif/pipelines/core/parsers/ParsedField.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * Models a parsed field.
+ * <p>
+ * A field can be as simple as a single {@link org.gbif.dwc.terms.DwcTerm} or a wrapper with several fields inside.
+ */
+public class ParsedField<T> {
+
+  // TODO: do we need confidence??
+
+  private T result;
+  private List<InterpretationIssue> issues = new ArrayList<>();
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/ParsedField.java]: new ArrayList<>(); can be removed and all fields must be 'final'

File: core/src/main/java/org/gbif/pipelines/core/parsers/ParsedField.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  private boolean successful;
+
+  private ParsedField(Builder<T> builder) {
+    this.result = builder.result;
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/ParsedField.java]: this. can be removed in the next 3 lines

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Utilities for assisting in the parsing of latitude and longitude strings into Decimals.
+ */
+public class CoordinateParseUtils {
+
+  private final static String DMS = "\\s*(\\d{1,3})\\s*[°d ]"
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java]: change  all "final static" to "static final"

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+      } catch (IllegalArgumentException e) {
+        return ParsedField.fail(new InterpretationIssue(IssueType.COORDINATE_INVALID, DwcTerm.decimalLongitude));
+      }
+    }
+
+    return validateAndRound(lat, lng);
+  }
+
+  private static boolean inRange(double lat, double lon) {
+    if (Double.compare(lat, 90) <= 0
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java]: can be done in 1 line

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        && Double.compare(lat, -90) >= 0
+        && Double.compare(lon, 180) <= 0
+        && Double.compare(lon, -180) >= 0) {
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isLat(String direction) {
+    if ("NS".contains(direction.toUpperCase())) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java]: can be done in 1 line

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    // if lat is out of range, but in range of the lng,
+    // assume swapped coordinates.
+    // note that should we desire to trust the following records, we would need to clear the flag for the records to
+    // appear in
+    // search results and maps etc. however, this is logic decision, that goes above the capabilities of this method
+    if (Double.compare(lat, 90) > 0 || Double.compare(lat, -90) < 0) {
+      // try and swap
+      if (inRange(lon, lat)) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java]: merge this into the outer if

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        dir = coord.charAt(coord.length() - 1);
+        coord = coord.substring(0, coord.length() - 1);
+      }
+      // without the direction chuck it at the regex
+      Matcher m = DMS_SINGLE.matcher(coord);
+      if (m.find()) {
+        return coordFromMatcher(m, 1, 2, 3, String.valueOf(dir));
+      }
+    }
+    throw new IllegalArgumentException();
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CoordinateParseUtils.java]: let's add a message to the Exception

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CountryMaps.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+   country are allowed (e.g. AB,CD and AB,EF).
+   This *overrides* the provided country, and includes an issue.
+  */
+  private static final Map<Country, Set<Country>> CONFUSED_COUNTRIES = Maps.newHashMap();
+  // And this is the same, but without the issue — we aren't exactly following ISO, but we accept it.
+  private static final Map<Country, Set<Country>> EQUIVALENT_COUNTRIES = Maps.newHashMap();
+
+  static {
+
+    InputStream in = CountryMaps.class.getClassLoader().getResourceAsStream(CONFUSED_COUNTRY_FILE);
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/CountryMaps.java]: move this into try (InputStream in = CountryMaps.class.getClassLoader().getResourceAsStream(CONFUSED_COUNTRY_FILE);) {
}
you can also ignore the all the closing logic

File: core/src/main/java/org/gbif/pipelines/core/parsers/legacy/Wgs84Projection.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        double lon2 = dstPt[0];
+        // verify the datum shift is reasonable
+        if (Math.abs(lat - lat2) > SUSPICIOUS_SHIFT || Math.abs(lon - lon2) > SUSPICIOUS_SHIFT) {
+          issues.add(new InterpretationIssue(IssueType.COORDINATE_REPROJECTION_SUSPICIOUS, DwcTerm.geodeticDatum));
+          LOG.debug("Found suspicious shift for datum={} and lat/lon={}/{} so returning failure and keeping orig coord",
+            datum, lat, lon);
+          return ParsedField.fail(new LatLng(lat, lon), issues);
+        }
+        // flag the record if coords actually changed
+        if (lat != lat2 || lon != lon2) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/legacy/Wgs84Projection.java]: this must be changed to (Double.compare(lat, lat2) + Double.compare(lon, lon2)) != 0 or something similar, compare 2 doubles in Java is vert unsafe: https://www.krenger.ch/blog/java-comparing-floating-point-numbers/

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/CoordinatesFunction.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    private static final Function<LatLng, LatLng> NEGATED_COORDS_FN =
+      latLng -> new LatLng(-1 * latLng.getLat(), -1 * latLng.getLng());
+    private static final Function<LatLng, LatLng> SWAPPED_COORDS_FN =
+      latLng -> new LatLng(latLng.getLng(), latLng.getLat());
+  }
+
+  public static List<IssueType> getIssueTypes(CoordinatesFunction transformation) {
+    if (transformation == PRESUMED_NEGATED_LAT) {
+      return Collections.singletonList(IssueType.PRESUMED_NEGATED_LATITUDE);
+    } else if (transformation == PRESUMED_NEGATED_LNG) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/CoordinatesFunction.java]: of 'elses' can be removed

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/CoordinatesValidator.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  /**
+   * Checks if the country and latitude belongs to Antarctica.
+   * Rule: country must be Country.ANTARCTICA or null and
+   * latitude must be less than (south of) {@link #ANTARCTICA_LATITUDE}
+   * but not less than -90°.
+   */
+  public static boolean isAntarctica(Double latitude, Country country) {
+    if (latitude == null) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/CoordinatesValidator.java]: all in 1 line?

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  private Optional<Country> containsAnyCountry(Collection<Country> possibilities, Collection<Country> countries) {
+    if (possibilities != null) {
+      return possibilities.stream().filter(countries::contains).findFirst();
+    }
+    return Optional.empty();
+  }
+
+  private List<Term> getCountryAndCoordinatesTerms() {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: static

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+  private boolean countriesFound(Optional<List<Country>> countriesOpt) {
+    return countriesOpt.filter(countries -> !countries.isEmpty()).isPresent();
+  }
+
+  private Optional<Country> matchWithCountry(List<Country> countries) {
+    return countries.contains(country) ? Optional.ofNullable(country) : Optional.empty();
+  }
+
+  private Optional<Country> containsAnyCountry(Collection<Country> possibilities, Collection<Country> countries) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: static, and 1 line?

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    if ((response.getBody() == null || response.getBody().isEmpty()) && isAntarctica(latLng.getLat(), country)) {
+      return Optional.of(Collections.singletonList(Country.ANTARCTICA));
+    }
+
+    return Optional.ofNullable(response.getBody());
+  }
+
+  private boolean countriesFound(Optional<List<Country>> countriesOpt) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: do not user Optional as params and the method is always negated

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    countryMatch = containsAnyCountry(CountryMaps.confused(country), countries);
+    if (countryMatch.isPresent()) {
+      issues.add(new InterpretationIssue(IssueType.COUNTRY_DERIVED_FROM_COORDINATES, DwcTerm.country));
+      return countryMatch;
+    }
+
+    return Optional.empty();
+  }
+
+  private Optional<List<Country>> getCountriesFromCoordinates(LatLng latLng) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: this method can be changed to return Optional.empty() when the list of countries in the response is empty, or we can remove the the usage of Optionals and user a List, int that way we can make the code more compact and probably remove the method 'countriesFound'

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    if (countryMatch.isPresent()) {
+      // country found
+      if (country == null) {
+        issues.add(new InterpretationIssue(IssueType.COUNTRY_DERIVED_FROM_COORDINATES, DwcTerm.country));
+      }
+      return ParsedField.success(ParsedLocation.newBuilder().country(countryMatch.get()).latLng(latLng).build(),
+                                 issues);
+    }
+
+    if (this.country != null) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: remove 'this.'

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.gbif.pipelines.core.parsers.location.CoordinatesValidator.isAntarctica;
+
+/**
+ * Matches the location fields related to Country and Coordinates to find possible mismatches.
+ */
+public class LocationMatcher {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java]: The logic of this class is hard to follow, can we split in 2 flows? parse when a country is specified and when it is not?

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  ) {
+    if (!countryParsed.isSuccessful()) {
+      LOG.info("country parsing failed");
+      issues.addAll(countryParsed.getIssues());
+    }
+
+    return Optional.ofNullable(countryParsed.getResult());
+  }
+
+  private static Country getFinalCountryMatch(Optional<Country> countryName, Optional<Country> countryCode) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java]: return countryCode.orElseGet(() -> countryName.orElse(null)); ?

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+    issues.addAll(coordsParsed.getIssues());
+
+    // return if coordinates parsing failed
+    if (!coordsParsed.isSuccessful()) {
+      LOG.info("coordinates parsing failed");
+      ParsedLocation parsedLocation = ParsedLocation.newBuilder().country(countryMatched).build();
+      return ParsedField.fail(parsedLocation, issues);
+    }
+
+    // set current parsed valuies
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java]: spelling mistake 'value'

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+    ParsedLocation parsedLocation =
+      ParsedLocation.newBuilder().country(countryMatched).latLng(coordsParsed.getResult()).build();
+
+    // if the coords parsing was succesfull we try to do a country match with the coordinates
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java]: successful

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  private static boolean isParsingSuccessful(Country countryMatched, ParsedField<ParsedLocation> match) {
+    return match.isSuccessful() && countryMatched != null;
+  }
+
+  private static void checkCountryMismatch(
+    List<InterpretationIssue> issues, Optional<Country> countryName, Optional<Country> countryCode
+  ) {
+    if (countryName.isPresent() && countryCode.isPresent() && !countryName.get().equals(countryCode.get())) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java]: I think that Optional.equals() can replace this expression: countryName.get().equals(countryCode.get())

File: core/src/main/java/org/gbif/pipelines/core/parsers/location/ParsedLocation.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Models a parsed location.
+ */
+public class ParsedLocation {
+
+  private final Country country;
+  private final LatLng latLng;
+
+  private ParsedLocation(Builder builder) {
+    this.country = builder.country;
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/parsers/location/ParsedLocation.java]: remove "this."

File: core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+
+  private Map<String, String> createParamsMap(LatLng latLng) {
+    Map<String, String> params = new HashMap<>();
+    params.put("lat", String.valueOf(latLng.getLat()));
+    params.put("lng", String.valueOf(latLng.getLat()));
+
+    return params;
+  }
+
+  protected Call<Collection<GeocodeResponse>> getCall(Map<String, String> params) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java]: missing @Override

File: core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    return GeocodeServiceRest.getInstance().getService().reverse(params);
+  }
+
+  protected String getErrorMessage() {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java]: @Override

File: core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    return "Error calling the geocode reverse WS";
+  }
+
+  protected List<Country> parseResponse(Collection<GeocodeResponse> response) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java]: @Override

File: core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+  public HttpResponse<List<Country>> getCountriesFromLatLng(LatLng latLng) {
+    if (!isInRange(latLng)) {
+      return HttpResponse.fail("lat and lng out of range", HttpResponse.ErrorCode.ABORTED);
+    }
+
+    return performCall(createParamsMap(latLng));
+  }
+
+  private Map<String, String> createParamsMap(LatLng latLng) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceClient.java]: static?

File: core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceRest.java [hunk 28]
```diff
@@ -1,9 +1,9 @@
...
+package org.gbif.pipelines.core.ws.client.geocode;
 
-import org.gbif.pipelines.ws.HttpClientFactory;
-import org.gbif.pipelines.ws.HttpConfigFactory;
-import org.gbif.pipelines.ws.config.Config;
-import org.gbif.pipelines.ws.config.Service;
+import org.gbif.pipelines.core.ws.HttpClientFactory;
+import org.gbif.pipelines.core.ws.HttpConfigFactory;
+import org.gbif.pipelines.core.ws.config.Config;
+import org.gbif.pipelines.core.ws.config.Service;
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/geocode/GeocodeServiceRest.java]: private final GeocodeService service;?

File: core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java [hunk 29]
```diff
@@ -63,26 +66,22 @@ private SpeciesMatchv2Client() {}
...
-    try {
-      Response<NameUsageMatch2> response = call.execute();
+    return performCall(params);
+  }
 
-      if (!response.isSuccessful()) {
-        String errorMessage = "Call to species match name WS failed: " + response.message();
-        return HttpResponse.fail(response.body(), response.code(), errorMessage, ErrorCode.CALL_FAILED);
-      }
+  protected Call<NameUsageMatch2> getCall(Map<String, String> params) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java]: @Override

File: core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java [hunk 30]
```diff
@@ -63,26 +66,22 @@ private SpeciesMatchv2Client() {}
...
+    return SpeciesMatchv2ServiceRest.getInstance().getService().match(params);
+  }
 
-      return HttpResponse.success(response.body());
-    } catch (IOException e) {
-      LOG.error("Error calling the match species name WS", e);
-      String errorMessage = "Error calling the match species name WS";
-      return HttpResponse.fail(null, errorMessage, ErrorCode.UNEXPECTED_ERROR);
-    }
+  protected String getErrorMessage() {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java]: @Override

File: core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java [hunk 31]
```diff
@@ -63,26 +66,22 @@ private SpeciesMatchv2Client() {}
...
+    return "Call to species match name WS failed";
+  }
 
+  protected NameUsageMatch2 parseResponse(NameUsageMatch2 body) {
```
Comment (fmendezh) [inline core/src/main/java/org/gbif/pipelines/core/ws/client/match2/SpeciesMatchv2Client.java]: @Override

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+
+    LatLng coordsCanada = new LatLng(LATITUDE_CANADA, LONGITUDE_CANADA);
+
+    ParsedField<ParsedLocation> match = LocationMatcher.newMatcher(coordsCanada).applyMatch();
+
+    Country canada = Country.CANADA;
+    Assert.assertEquals(canada, match.getResult().getCountry());
+    Assert.assertEquals(coordsCanada, match.getResult().getLatLng());
+    Assert.assertTrue(match.isSuccessful());
+    Assert.assertTrue(match.getIssues().get(0).getIssueType().equals(IssueType.COUNTRY_DERIVED_FROM_COORDINATES));
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java]: user == to compare enums

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java [hunk 33]
```diff
@@ -0,0 +1,275 @@
...
+
+    ParsedField<ParsedLocation> match = LocationMatcher.newMatcher(wrongCoords)
+      .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_LAT)
+      .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_LNG)
+      .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_COORDS)
+      .addAdditionalTransform(CoordinatesFunction.PRESUMED_SWAPPED_COORDS)
+      .applyMatch();
+
+    Assert.assertFalse(match.isSuccessful());
+    Assert.assertTrue(match.getIssues().get(0).getIssueType().equals(IssueType.COUNTRY_COORDINATE_MISMATCH));
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java]: user == to compare enums

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java [hunk 34]
```diff
@@ -0,0 +1,275 @@
...
+  public void whenMatchConfusedCountryThenReturnConfused() throws IOException {
+    enqueueResponse(GREENLAND_REVERSE_RESPONSE);
+
+    LatLng coords = new LatLng(71.7, -42.6);
+    ParsedField<ParsedLocation> match = LocationMatcher.newMatcher(coords, Country.DENMARK).applyMatch();
+
+    Assert.assertEquals(Country.GREENLAND, match.getResult().getCountry());
+    Assert.assertEquals(coords, match.getResult().getLatLng());
+    Assert.assertTrue(match.isSuccessful());
+    Assert.assertTrue(match.getIssues().get(0).getIssueType().equals(IssueType.COUNTRY_DERIVED_FROM_COORDINATES));
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java]: ==

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java [hunk 35]
```diff
@@ -0,0 +1,275 @@
...
+  public void emptyCoordinates() {
+    LocationMatcher.newMatcher(new LatLng()).applyMatch();
+  }
+
+  @Test
+  public void outOfRangeCoordinates() {
+    ParsedField<ParsedLocation> match = LocationMatcher.newMatcher(new LatLng(200, 200)).applyMatch();
+
+    Assert.assertFalse(match.isSuccessful());
+    Assert.assertTrue(match.getIssues().get(0).getIssueType().equals(IssueType.COUNTRY_COORDINATE_MISMATCH));
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationMatcherTest.java]: ==

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java [hunk 36]
```diff
@@ -0,0 +1,141 @@
...
+                        .contains(IssueType.COORDINATE_INVALID));
+  }
+
+  @Test
+  public void givenInvalidCountryWhenParsedThenReturnIssue() {
+    // only with country
+    ExtendedRecord extendedRecord = new ExtendedRecordCustomBuilder().id(TEST_ID).country("foo").build();
+    ParsedField<ParsedLocation> result = LocationParser.parseCountryAndCoordinates(extendedRecord);
+    Assert.assertFalse(result.isSuccessful());
+    Assert.assertEquals(null, result.getResult().getCountry());
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java]: assertNull

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java [hunk 37]
```diff
@@ -0,0 +1,141 @@
...
+                        .contains(IssueType.COUNTRY_INVALID));
+  }
+
+  @Test
+  public void givenInvalidCountryCodeWhenParsedThenReturnIssue() {
+    // only with country code
+    ExtendedRecord extendedRecord = new ExtendedRecordCustomBuilder().id(TEST_ID).countryCode("foo").build();
+    ParsedField<ParsedLocation> result = LocationParser.parseCountryAndCoordinates(extendedRecord);
+    Assert.assertFalse(result.isSuccessful());
+    Assert.assertEquals(null, result.getResult().getCountry());
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java]: assertNull

File: core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java [hunk 38]
```diff
@@ -0,0 +1,141 @@
...
+  public void givenOnlyCoordsWhenParsedThenReturnCoordsWithDerivedCountry() throws IOException {
+    enqueueResponse(CHINA_REVERSE_RESPONSE);
+
+    // only with coords
+    ExtendedRecord extendedRecord =
+      new ExtendedRecordCustomBuilder().id(TEST_ID).decimalLatitude("30.2").decimalLongitude("100.2344349").build();
+    ParsedField<ParsedLocation> result = LocationParser.parseCountryAndCoordinates(extendedRecord);
+    Assert.assertFalse(result.isSuccessful());
+    Assert.assertEquals(Country.CHINA, result.getResult().getCountry());
+    Assert.assertEquals(30.2d, result.getResult().getLatLng().getLat().doubleValue(), 0);
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/parsers/location/LocationParserTest.java]: i think that doubleValue is not needed

File: core/src/test/java/org/gbif/pipelines/core/ws/config/HttpConfigFactoryTest.java [hunk 39]
```diff
@@ -1,6 +1,8 @@
-package org.gbif.pipelines.ws.config;
+package org.gbif.pipelines.core.ws.config;
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/ws/config/HttpConfigFactoryTest.java]: replace concat by  + in line 27

File: core/src/test/java/org/gbif/pipelines/core/ws/MockServer.java [hunk 40]
```diff
@@ -1,4 +1,4 @@
-package org.gbif.pipelines.ws;
+package org.gbif.pipelines.core.ws;
```
Comment (fmendezh) [inline core/src/test/java/org/gbif/pipelines/core/ws/MockServer.java]: MockServer can be wrapped in a  @ClassRule and into a org.junit.rules.ExternalResource

File: core/src/main/java/org/gbif/pipelines/core/interpretation/LocationInterpreter.java (hunk 1 above)
Comment (asturcon) [reply core/src/main/java/org/gbif/pipelines/core/interpretation/LocationInterpreter.java]: fixed. That different indentation was done with our common formatter. If it's wrong I think we should fix it in order to be able to apply it everywhere.

**Comment by asturcon** [comment]
@fmendezh all changes applied
